### DISCUSSION TITLE
Use Fastapi parameter lifespan to load middleware lifespan is decorated by asynccontextmanager scheme

### DISCUSSION THREAD

**Discussion by pcorvoh** [fastapi/fastapi#9357] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware
from contextlib import asynccontextmanager
from uvicorn import run


@asynccontextmanager
async def add_middleware(app: FastAPI, ) -> None:
    app.add_middleware(
            CORSMiddleware,
            allow_origins=["*"],
            allow_credentials=True,
            allow_methods=["*"],
            allow_headers=["*"],
    )
    yield


app = FastAPI(lifespan=add_middleware)


@app.get("/do_long_thing")
async def do_long_thing():
    return {"msg": "hello,world"}

-----------------------I am a beautiful dividing line-----------------------
# Here is the second way to write the loading middleware
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware
from contextlib import asynccontextmanager
from uvicorn import run

app = FastAPI()


@app.on_event("startup")
async def add_middleware() -> None:
    app.add_middleware(
            CORSMiddleware,
            allow_origins=["*"],
            allow_credentials=True,
            allow_methods=["*"],
            allow_headers=["*"],
    )


@app.get("/do_long_thing")
async def do_long_thing():
    return {"msg": "hello,world"}
```

### Description

When I run this code after I upgrade the version, I am prompted with `RuntimeError: Cannot add middleware after an application has started` and cannot start fastapi's service, as if fastapi has done a null processing
### Demand: How do I need to use a more elegant way of middleware loading in a high version
Here is the complete error stack information：
```
INFO:     Will watch for changes in these directories: ['E:\\softtest_python_scripts\\learnfastapi']
INFO:     Uvicorn running on http://127.0.0.1:9999 (Press CTRL+C to quit)
INFO:     Started reloader process [7708] using statreload
WARNING:  The --reload flag should not be used in production on Windows.
INFO:     Started server process [6116]
INFO:     Waiting for application startup.
ERROR:    Traceback (most recent call last):
  File "E:\soft\Python3.9\lib\site-packages\starlette\routing.py", line 677, in lifespan
    async with self.lifespan_context(app) as maybe_state:
  File "E:\soft\Python3.9\lib\contextlib.py", line 181, in __aenter__
    return await self.gen.__anext__()
  File "E:\softtest_python_scripts\learnfastapi\fastapi_platform\middleware_fastapi.py", line 10, in add_middleware
    app.add_middleware(
  File "E:\soft\Python3.9\lib\site-packages\starlette\applications.py", line 139, in add_middleware
    raise RuntimeError("Cannot add middleware after an application has started")
RuntimeError: Cannot add middleware after an application has started
```

### Operating System

Windows

### FastAPI Version

0.95.0

### Python Version

Python 3.9.9

**Comment by Kludex** [comment, marked as answer]
That's a side effect from: https://github.com/encode/starlette/pull/2017

On your case, you should do as following:
```python
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware

app = FastAPI()
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

@app.get("/do_long_thing")
async def do_long_thing():
    return {"msg": "hello,world"}
```

**Comment by pcorvoh** [comment]
Thank you very much for your answer, this is probably my best solution, but since I currently have a lot of middleware and I was using asynchronous for loading before, it looks like I need to make my code elegant again
